README.md
# Yui - personal task manager
Git based personal task manager.
100% command line.
(also, Japanese girl name)

> According to TDD as described in "Extreme Programming: Test Driven Development" by Kent Beck,
we need to start every day with blank sheet of paper, white down small tasks
and stroke out what is done. This is digital replacement for that paper sheets.

## Install
### Pip 3
```
pip3 install yui
```


## 4 dummies
 0. `yui` without arguments will display help

Workflow:
 1. Add task to the heap `yui create task name or short descriprion`
    1. Show tasks in heap `yui list heap`
 2. Pick the task to current day schedule `yui pick %taskId%` 
    1. Show tasks for current day `yui list cur`
 3. Open task in text editor  `yui open %taskId%`
    1. Task file is just md file with yaml header
    2. Adjust status in header
    3. Write any notes below
    4. Save
 4. When you done working, or before next day
    1. Move unfinished tasks back to heap `yui reset all`
    2. Archive tasks with status *done*  `yui archive today`
 5. View archive for specific date `yui list 2023-01-11`
 6. Run manual git command on task list `yui git %git_command% %git_command_args%`
 7. Adjust visible scope with `yui scope`
 
Optional configuration:
`.yui/config.yaml`
`.yui/projects/%projectName%.yaml`
 
## How it works
 1. There is git repository behind the scenes. So you have history and you can sync tasks using any git server. History, branches, etc.
    1. Create/pick/reset/archive commands will invoke `git add` + `git commit` commands automatically.
 2. Task data stored in plain text files.
    1. Format markdown(.md) with yaml header.
    2. You work with single task using plain text editor, like kate
 3. yui tool is used to organize and navigate .md files
 
## The method
### Step 1: Write it down and forget. 
Once you spot new task or idea - you just add it to the heap, and continue with your current work. So you stay focused.
```
yui create
```
*Heap* is the most chaotic, unorganised, unsorted, *backlog* you can imagine.
Do not waste your time on details, just stock pile it in the heap as is.

Imagine that you are working with paper stickers and you have a big box of chaotic notes written on stickers - that's the heap.
```
yui list heap
```
Will show you tasks in the heap in form of table
```
 id | context  | project    | name                                                                   | status
----|----------|------------|------------------------------------------------------------------------|-------
 7  | personal | yui        | show creation date column in task list                                 | new   
 15 | personal | yui        | sanitize slashes in task filename                                      | new
```
First column is *taskId* you will need it to manipulate the task

### Step 2: pick the task into daily plan
Pick all tasks you are planning to work with today. 
```
yui pick %taskId%
```
Will pick the task
```
yui list cur
```
Will display tasks for current day.

If you made a mistake, you can return the task back to heap
```
yui reset %taskId%
```

### Step 3: open the task in text editor
```
yui open %taskId%
```
### Step 4: change status to *work*
Adjust yaml header, simple replace `status: new` with `status: work`.

You can use any custom statuses, but buildin - *new*, *work*, *done* will be highlighted in `yui list` output.

### Step 5: Make notes while you are working on the task
As for it's just .md file, you can make any notes behind yaml header.

### Note: you can work with as many tasks at once as you want
That's just plain text files, after all.

### Step 6: change task status to *done*
In yaml header, replace `status: work` with `status: work`.

### Step 7: check your progress
```
yui list cur
```
Will get more green lines while you complete the tasks.

### Step 8: Cleanup workspace
```
yui reset all
```
Will return unfinished tasks back to the heap
```
yui archive today
```
Will move only *done* tasks to archive folder with current day. You can also use "yesterday" or specify date as "YYYY-MM-DD", or anything else recognized by php `strtotime()` function.

To view archived tasks use:
```
yui list YYYY-MM-DD
```

### Step 9: apply git
All git commands are mapped with `yui git`.

Most used:
    - `yui git log` history of changes in tasks
    - `yui git remote add origin %link%` link your task list with remote repository
    - `yui git push` save local changes to remote repository
    - `yui git pull` load fresh changes from remote repository
   
## Apllication working folder
Defaut:
 - Linux: `~/.yui`
 - Windows: `c:\Userts\%username%\AppData\yui`
Can be overwriten with `YUI_HOME` environment variable, like `YUI_HOME=mypath yui list heap`


## config.yaml
> Location: %application-working-folder%/config.yaml

Example:
```yaml
---
# % will be replaced with filename
# Examples:
#   To run gui editor: nohup kate % > /dev/null 2>&1 &
#   To run in terminal: mcedit %
# You can use nano, vi, mcedit etc
# If not specified, tsk will try EDITOR environment variable and then mcedit, nano, vim, vi, ee in that order
editor: nohup kate % > /dev/null 2>&1 &
---
```

## Project configuration
> Location: %application-working-folder%/projects/%projectName%.yaml

Example file:projects/myProject.yaml:
```yaml
---
# this will make YUI_PROJECT=myProject yui create test
# work the same way as YUI_PROJECT=myProject YUI_CONTEXT=myContext yui create test
defaultContext: myContext
---
```

## View archives
```
yui list-archives
```
Will display list of existing archives, from application working folder

To view tasks in single archive use
```
yui list %archiveName%
```

## Advanced options
You can pick range of items like so `pick 141,142,143..154`


CHANGELOG.md
# CHANGELOG

## [2023-02-20]
### Added
 - Macos and windows changes for yui open
 - error handling for random string instead of taskId
 - Open with notepad on windows, if editor not specified

## [2023-02-15]
### Added
 - list-archives command

## [2023-02-13]
### Fixed
 - Yaml dependency problem, etc
 - No linux colors in windows console (will be just white)

## [2023-02-12]
### Fixed
 - Default folder on windows was %userfolder%/yui, should be %userfolder%/AppData/Local, fixed. 

## [2023-02-05]
### Added
 - check if git exist, disable automatic commits if not, display warning

## [2023-02-04]
### Added
 - pick range of items like so: `pick 141,142,143..154`

## [2023-02-03]
### Added
 - Hint for next possible commands under the list of tasks, different for cur and heap

## [2023-02-02]
### Added
 - Order numbers ( "No." column on the left )
 - pick/reset/open/drop tasks by order number ( by id still works, and it is default behaviour )
    as for order number is different for `list heap` and `list cur` you need to prefix order number with container name, example: `open cur3`
 - table title in bottom-right corner


src/yui/tsklib.py
#!/usr/bin/env python
import os, glob, yaml, datetime, subprocess
from appdata import AppDataPaths
from sanitize_filename import sanitize
from pathlib import Path



Title="Yui"
cmd="yui"

def tskpath():
    default=cmd
    home = os.getenv("YUI_HOME", default)
    path = AppDataPaths(home).app_data_path
    os.makedirs(path, exist_ok=True)
    return path
    pass

def getTaskFilenameById(id, location="*"):
    files = findTaskFiles( location, "*."+id+".md")
    if len(files) == 0:
        suffix = ""
        if location !="*":
            suffix = " in "+location
        print("task with id="+id+" not found"+suffix)
        exit(1)
    pass;
    if len(files) != 1:
        print("more than one task with id="+id+" found. That should never happen")
        exit(1)
    pass;
    return files[0]
    pass

def getTaskFilenameByIdOrNum( idOrNum, location="*"):
    if idOrNum.isnumeric():
        return getTaskFilenameById( idOrNum, location )
        pass
    if idOrNum[:2] == "id":
        return getTaskFilenameById( idOrNum[2:], location )
        pass
    if idOrNum[:3] == "cur":
        return getTaskFilenameByNum( idOrNum[3:], "cur")
        pass
    if idOrNum[:4] == "heap":
        return getTaskFilenameByNum( idOrNum[4:], "heap")
        pass
    print("Unknown task id or number format");
    pass

def getTaskByIdOrNum( idOrNum, location="*"):
    if idOrNum.isnumeric():
        return getTaskById( idOrNum, location )
        pass
    if idOrNum[:2] == "id":
        return getTaskById( idOrNum[2:], location )
        pass
    if idOrNum[:3] == "cur":
        return getTaskByNum( idOrNum[3:], "cur")
        pass
    if idOrNum[:4] == "heap":
        return getTaskByNum( idOrNum[4:], "heap")
        pass
    print("Unknown task id or number format");
    pass



def getTaskFilenameByNum( num, location ):
    return getTaskByNum( num, location )["fullfilename"]
    pass

def getTaskById(id, location):
    filename = getTaskFilenameById( id, location )
    task = loadYaml( filename )
    task["fullfilename"] = filename
    return task
    pass

def getTaskByNum( num, location ):
    tasks = listTasks( location )
    for task in tasks:
        if int(task["No."]) == int(num):
            return task
    print("Task "+num+" not found in "+location)
    pass



def findTaskFiles(location, pattern):
    return glob.glob( tskpath() + "/"+location+"/*/"+pattern)
    pass

def listTasks(location):
    scope = getScope();
    files = findTaskFiles( location, "*.md" )
    tasks = []
    for filename in files:
        task = loadYaml(filename)
        task["fullfilename"] = filename
        addItem = True
        for key in scope:
            addItem = addItem and ( scope[key] == "" or scope[key] == task[key] )
            pass
        if not addItem:
            continue
            pass
        tasks.append( task )
        pass
    tasks = sorted(tasks, key = lambda task : task["id"])
    key = 0
    for task in tasks:
        key = key+1
        task["No."] = key
        pass
    return tasks
    pass

def getConfigParam(param):
    return loadYaml( tskpath() + "/config.yaml" )[param]
    pass

def loadYaml(filename):
    with open( filename ) as f:
        return next(yaml.load_all(f, Loader=yaml.loader.UnsafeLoader))
        pass
    pass

def saveYaml(filename, data):
    with open( filename, 'w' ) as f:
        yaml.dump( data, f )
        pass
    pass

def gitExist()->bool:
    try:
        re = subprocess.run(["git","-v"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    except:
        return False
    return re.returncode == 0
    pass

def gitInitIfNotPresent():
    workingdir = tskpath()
    if os.path.isdir( workingdir+"/.git"):
        return
        pass
    curpath = os.getcwd();
    os.chdir(workingdir);
    os.system("git init > /dev/null");
    os.chdir(curpath);
    pass

def gitAddCommitTask(message):
    if not gitExist():
        print("git not exist");
        return
        pass
    gitInitIfNotPresent()
    curpath = os.getcwd();
    os.chdir(tskpath());
    os.system("git add ./*.md > /dev/null && git commit -m '"+message+"' > /dev/null");
    os.chdir(curpath);
    pass

def getLastId():
    files = findTaskFiles("*", "*.md")
    maxId = -1
    for filename in files :
        # Open the file and load the file
        id = loadYaml( filename )["id"]
        if id > maxId:
            maxId = id
            pass
        pass
    return maxId
    pass

def newScope():
    return {
        "project":"",
        "context":""
        }
    pass

def loadScope():
    scope = newScope()
    try:
        data = loadYaml( tskpath()+"/scope.yaml")
        for key in scope:
            try:
                scope[key] = data[key]
            except:
                pass
            pass
    except:
        pass
    return scope
    pass

def mergeScope( orig, changes ):
    scope = orig
    for key in changes:
        if changes[key] == "":
            continue
        scope[key] = changes[key]
    return scope
    pass

def parseFilter( filterstring ):
    scope = newScope()
    if filterstring == "":
        return scope
        pass
    # split by ,
    pairs = filterstring.split(",")
    for pair in pairs:
        keyvalue = pair.split(":")
        key = keyvalue[0]
        value = keyvalue[1]
        if key == "c":
            key = "context"
            pass
        if key == "p":
            key = "project"
            pass
        scope[key] = value
        pass
    return scope
    pass

def getDefaultContextForProject( projectName ):
    try:
        return loadYaml( tskpath() + "/projects/" + projectName + ".yaml")["defaultContext"]
    except:
        return ""
    pass

def getScope():
    scope = loadScope()
    scope = mergeScope( scope, parseFilter(os.getenv(cmd.upper(),"")) )
    for key in scope:
        scope[key] = os.getenv(cmd.upper() + "_" + key.upper(), scope[key])
        pass
    if scope["context"] == "" and scope["project"] != "":
        scope["context"] = getDefaultContextForProject( scope["project"] )
        pass
    return scope
    pass

def saveScope( scope ):
    saveYaml( tskpath()+"/scope.yaml", scope)
    pass

def createTask( name ):
    name = name.replace("\"","'")
    tasknameArr = name.split(" ")
    taskname = '_'.join( tasknameArr )
    taskname = sanitize( taskname ).replace("'","").replace("\"","").replace("`","")
    taskDatetime = datetime.datetime.today()
    id = str( int( getLastId() )+1 )
    path = tskpath() + "/heap/new"
    filename = taskDatetime.strftime("%Y-%m-%d_%H.%M.%S_%z_")+taskname+"."+id+".md"
    os.makedirs(path, exist_ok=True)

    scope = getScope()

    Path( path + "/" + filename ).write_text("""---
name: \""""+name+"""\"
created: """+taskDatetime.strftime("%Y-%m-%d %H:%M:%S %z")+"""
context: """+scope["context"]+"""
project: """+scope["project"]+"""
filename:  """+filename+"""
status: new
id: """+id+"""
---
""", encoding='utf-8')
    
    gitAddCommitTask("created "+id);
    pass

def pickTask(id):
    filename = getTaskFilenameByIdOrNum(id, "heap")
    task = loadYaml(filename)
    targetPath = tskpath() + "/cur/" + task["status"]
    os.makedirs(targetPath, exist_ok=True)
    os.rename( filename, targetPath + "/" + task["filename"]);
    gitAddCommitTask("pick "+id);
    pass

def rangeQueryToArray(query):
    tasks = listTasks("heap")
    taskIds = []
    if ".." in query:
        for task in tasks:
            taskIds.append( task["id"] )
            pass
        pass
    #print(taskIds)
    ids = []
    queryElements = query.split(",")
    for queryElement in queryElements:
        taskRange = queryElement.split("..")
        if len(taskRange) == 1:
            ids.append( queryElement )
            continue
        pass
        if len(taskRange) > 2:
            raise Exception("Mailformed id range query")
        pass
        for item in range(int(taskRange[0]), int(taskRange[1])+1): # TODO generalize for curNNN, heapNNN
            if item in taskIds:
                ids.append(str(item))
                pass
            pass
        pass
    return ids
    pass

def pickTasks(query):
    '''
    pick one or more tasks using array and range, like pick 141,142,143..150
    pick only from results of listTasks(heap)
    '''
    items = rangeQueryToArray(query)
    for item in items:
        #print(item)
        pickTask(item)
        pass
    pass
